### PR TITLE
[onert-micro] Introduce OMTraining entities

### PR DESCRIPTION
**Pull request by BalyshevArtem**
This pr introduces OMTrainingInterpreter, OMTrainingContext and OMTrainingRuntimeModule entities.

for issue #12873
from draft: #13107

ONE-DCO-1.0-Signed-off-by: Artem Balyshev <a.balyshev@samsung.com>

### CODE REVIEW

File: onert-micro/onert-micro/include/OMConfig.h [hunk 1]
```diff
@@ -17,13 +17,86 @@
...
+ * epsilon - used by ADAM optimizer
+ * num_Step - used by ADAM optimizer
+ */
+struct OMTrainingContext
+{
+  uint32_t batch_size = 1;
+  uint32_t num_of_train_layers = 0;
+  OMTrainOptimizer optimizer = SGD;
+  OMLoss loss = MSE;
+  float learning_rate = 0.f;
```
Comment (chunseoklee) [inline onert-micro/onert-micro/include/OMConfig.h:75]: how about default value other than 0

File: onert-micro/onert-micro/include/OMConfig.h [hunk 2]
```diff
@@ -17,13 +17,86 @@
...
+  float beta = 0.9f;
+  float beta_squares = 0.9f;
+  float epsilon = 10e-8;
+  uint32_t num_step = batch_size;
```
Comment (chunseoklee) [inline onert-micro/onert-micro/include/OMConfig.h:79]: since this is for checkpoint, this should be 0 at start. It is better to increase this after each step internally. But, this value is not maintained inside interpreter. Thus, I try to increase this on nnfw api level like : https://github.com/Samsung/ONE/commit/89af72f45123f16f50f9efea865d9a9a9c83cfe5

File: onert-micro/onert-micro/src/OMTrainingInterpreter.cpp [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+  if (not out_file.is_open())
+    return UnknownError;
+
+  // Write data
+  out_file.write(config.model_ptr, config.model_size);
+
+  // Close file
+  out_file.close();
+#else
+  assert(fasle && "Not supported");
```
Comment (chunseoklee) [inline onert-micro/onert-micro/src/OMTrainingInterpreter.cpp:69]:
```suggestion
  assert(false && "Not supported");
```

File: onert-micro/onert-micro/src/core/OMTrainingRuntimeModule.cpp [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+ *  Train single step: run forward graph (with data which was set in SetInput) ->
+ *  -> calculate error (with target data which was set in SetTarget) ->
+ *  -> run backward graph -> update optimizer state -> after batch_size steps update weights
+ *  Warning: before using trainSingleStep call: 1) importTrainModel; 2) setInput; 3) setTarget
+ */
+OMStatus OMTrainingRuntimeModule::trainSingleStep(OMConfig &config)
+{
+  OMStatus status = Ok;
+  uint32_t batch_size = config.training_context.batch_size;
+  config.training_context.num_step += 1;
```
Comment (chunseoklee) [inline onert-micro/onert-micro/src/core/OMTrainingRuntimeModule.cpp:80]: Thank you.

File: onert-micro/onert-micro/include/OMConfig.h [hunk 5]
```diff
@@ -17,13 +17,86 @@
...
+enum OMTrainOptimizer
+{
+  SGD,
+  ADAM,
+};
+
+/*
+ * OMMetrics - enum to store metrics supported by training with onert-micro
+ */
+enum OMMetrics
```
Comment (chunseoklee) [inline onert-micro/onert-micro/include/OMConfig.h:38]: @BalyshevArtem Note that nnfw api does not expose Metric Evaluation. That is, nnfw api will expose loss evaluation function.

**Review by chunseoklee — DISMISSED** [review]
LGTM. Thank you

File: onert-micro/onert-micro/include/OMConfig.h (hunk 1 above)
Comment (BalyshevArtem) [reply onert-micro/onert-micro/include/OMConfig.h]: Done

File: onert-micro/onert-micro/include/OMConfig.h (hunk 2 above)
Comment (BalyshevArtem) [reply onert-micro/onert-micro/include/OMConfig.h]: Changed to 0

File: onert-micro/onert-micro/src/OMTrainingInterpreter.cpp (hunk 3 above)
Comment (BalyshevArtem) [reply onert-micro/onert-micro/src/OMTrainingInterpreter.cpp]: Thank you, fixed it

File: onert-micro/onert-micro/src/core/OMTrainingRuntimeModule.cpp (hunk 4 above)
Comment (BalyshevArtem) [reply onert-micro/onert-micro/src/core/OMTrainingRuntimeModule.cpp:80]: Only problem, where to reset this value to 0. For example we now start new epoch and that is why we need to reset this value again to 0.

Comment (chunseoklee) [reply onert-micro/onert-micro/src/core/OMTrainingRuntimeModule.cpp:80]: AFAIK, this num_step is used only for ADAM. Then it does not need to be reset at new epoch.

**Review by chunseoklee — APPROVED** [review]

File: onert-micro/onert-micro/src/OMTrainingInterpreter.cpp [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+#ifndef DIS_STREAM
+  // Open or create file
+  // Note: if the file existed, it will be overwritten
+  std::ofstream out_file(save_path, std::ios::binary | std::ios::trunc);
+  if (not out_file.is_open())
+    return UnknownError;
+
+  // Write data
+  out_file.write(config.model_ptr, config.model_size);
```
Comment (Torrero) [inline onert-micro/onert-micro/src/OMTrainingInterpreter.cpp:64]: Maybe it would be better to catch and to process some exceptions of the `write` function and in case of arising ones to return Error status (the same can be apply for `saveCheckpoint`). Just check `ios_base::badbit` state.

**Review by Torrero — APPROVED** [review]
LGTM

File: onert-micro/onert-micro/src/OMTrainingInterpreter.cpp (hunk 6 above)
Comment (BalyshevArtem) [reply onert-micro/onert-micro/src/OMTrainingInterpreter.cpp:64]: IN next pr will add this, thank you
